pyispyb/core/schemas/userportalsync.py

```diff
@@ -0,0 +1,143 @@
+from typing import List, Optional, Literal
+from pydantic import BaseModel, conlist, root_validator
+from pydantic_sqlalchemy import sqlalchemy_to_pydantic
+from pyispyb.core.schemas.laboratories import Laboratory
+from ispyb import models
+from datetime import datetime
+
+PydanticPerson = sqlalchemy_to_pydantic(
+    models.Person, exclude={"personId", "laboratoryId", "recordTimeStamp"}
+)
+PydanticProposal = sqlalchemy_to_pydantic(
+    models.Proposal, exclude={"proposalId", "personId", "bltimeStamp"}
+)
+PydanticSession = sqlalchemy_to_pydantic(
+    models.BLSession, exclude={"sessionId", "proposalId", "bltimeStamp", "comments"}
+)
+
+"""
+Excluding _global field to avoid a Pydantic RuntimeWarning
+RuntimeWarning: fields may not start with an underscore, ignoring "_global"
+"""
+PydanticProtein = sqlalchemy_to_pydantic(
+    models.Protein,
+    exclude={"proteinId", "proposalId", "personId", "bltimeStamp", "_global"},
+)
+PydanticSessionHasPerson = sqlalchemy_to_pydantic(
+    models.SessionHasPerson, exclude={"sessionId", "personId"}
+)
+PydanticLabContact = sqlalchemy_to_pydantic(
+    models.LabContact,
+    exclude={"labContactId", "proposalId", "personId", "recordTimeStamp"},
+)
+
+
+class Person(PydanticPerson):
+    # At least login or externalId required to be able to check for existing Person in DB (to update or create)
+    login: Optional[str] = None
+    externalId: Optional[int] = None
+
+    # https://github.com/samuelcolvin/pydantic/issues/506
+    @root_validator()
+    def check_login_or_externalId(cls, values):
+        if (values.get("login") is None) and (values.get("externalId") is None):
+            raise ValueError(
+                "either login or externalId is required for a Person entity"
+            )
+        return values
+
+    class Config:
+        orm_mode = True
+
+
+class PersonSessionOptions(PydanticSessionHasPerson):
+    role: Optional[
+        Literal[
+            "Local Contact",
+            "Local Contact 2",
+            "Staff",
+            "Team Leader",
+            "Co-Investigator",
+            "Principal Investigator",
+            "Alternate Contact",
+        ]
+    ]
+
+
+class PersonProposalLaboratory(Person):
+    laboratory: Laboratory
```

Comment:
Person.Laboratory is optional in the database. Maybe it should be here too?

Reply:
Yes, I guess it can be optional here too, I'll change it.